types/index.d.ts

```diff
@@ -0,0 +1,151 @@
+// TODO streams
+// TODO types
+
+export namespace Type {
+	type Schema = string;
+
+	interface Type {
+		// TODO clone(val, opts)
+		// TODO compare
+		// TODO compareBuffers(buf1, buf2)
+		// TODO createResolver(type, opts)
+		// TODO decode(buf, pos, resolver)
+		// TODO encode(val, buf, pos)
+		// TODO equals(type)
+		// TODO fingerprint(algorithm)
+		fromBuffer(buffer: Buffer, resolver: any, noCheck: boolean): Type; // TODO
+		// TODO fromString(str)
+		// TODO inspect()
+		// TODO isValid(val, opts)
+		// TODO random()
+		// TODO schema(opts)
+		toBuffer(value: object): Buffer;
+		// TODO toJSON()
+		// TODO toString(val)
+		// TODO wrap(val)
+	}
+
+	function forSchema(schema: Schema): Type;
+	function forValue(value: object): Type;
+	// TODO function forTypes(types, opts)
+	function isType(arg: any): boolean; // TODO
+}
+
+export namespace Service {
+	interface Protocol {
+	}
+
+	interface ClientChannel extends EventEmitter {
+		readonly client: Client;
+		readonly timeout: number;
+		readonly destroyed: boolean;
+		readonly draining: boolean;
+		readonly pending: number;
+		ping(timeout: number, cb: any): void;
+		destroy(noWait: boolean): void;
+	}
+
+	interface ServerChannel extends EventEmitter  {
+		readonly server: Server;
+		readonly destroyed: boolean;
+		readonly draining: boolean;
+		readonly pending: number;
+		destroy(noWait: boolean): void;
+	}
+
+	interface ClientOptions {
+		buffering: boolean;
+		cache: any;
+		channelPolicy: any;
+		strictTypes: boolean;
+		timeout: number;
+		remoteProtocols: boolean;
+	}
+
+	interface ServerOptions {
+		objectMode: boolean;
+	}
+
+	interface Service {
+		createClient(options?: ClientOptions): Client;
+		createServer(options?: ServerOptions): Server;
+		message(name: string): any;
+		type(name: string): any;
+		inspect(name: string): any;
+	}
+
+	type TransportFunction = () => void; // TODO
+
+	type Transport = stream.Duplex | TransportFunction;
+
+	interface ChannelCreateOptions {
+		objectMode: boolean;
+	}
+
+	interface ChannelDestroyOptions {
+		noWait: boolean;
+	}
+
+	interface Server extends EventEmitter {
+		readonly service: Service;
+		// on<message>()
+		activeChannels(): ServerChannel[];
+		createChannel(transport: Transport, options?: ChannelCreateOptions): ServerChannel;
+		onMessage<T>(name: string, handler: (arg1: any, callback: Callback<T>) => void): this;
+		remoteProtocols(): Protocol[];
+		use(...args: any[]): this;
+	}
+
+	interface Client extends EventEmitter {
+		activeChannels(): ClientChannel[];
+		createChannel(transport: Transport, options?: ChannelCreateOptions): ClientChannel;
+		destroyChannels(options?: ChannelDestroyOptions): void;
+		emitMessage<T>(name: string, req: any, options?: any, callback?: Callback<T>): void // TODO
+		remoteProtocols(): Protocol[];
+		use(...args: any[]): this;
+	}
+
+	function compatible(client: Client, server: Server): boolean;
+	function forProtocol(protocol: Protocol): Service;
```

Comment:
This might be too restrictive. It's sometimes useful to instantiate services from inlined JavaScript objects, which I think this would prevent, right?